### PR TITLE
Support definition based extraction for lists with Non - primitive types

### LINKED ISSUE #1769: Definition Based Extraction: Issue in extracting Observation value of type valueCoding
**Describe the bug**
Issue in extracting `Observation.value` for type **valueCoding** using Definition Based Extraction. Throws NullPointerException

**Component**
SDC library

**To Reproduce**
1. Use below json as questionnaire string.
```JSON
{
              "resourceType": "Questionnaire",
              "item": [
                {
                  "linkId": "9",
                  "type": "group",
                  "extension": [
                    {
                      "url": "http://hl7.org/fhir/uv/sdc/StructureDefinition/sdc-questionnaire-itemExtractionContext",
                      "valueExpression": {
                        "expression": "Observation"
                      }
                    }
                  ],
                  "item": [
                    {
                      "linkId": "9.1",
                      "type": "group",
                      "definition": "http://hl7.org/fhir/StructureDefinition/Observation#Observation.valueCodeableConcept",
                      "item" :[
                        {
                          "linkId": "9.1.1",
                          "type": "choice",
                          "definition": "http://hl7.org/fhir/StructureDefinition/Observation#Observation.valueCodeableConcept.coding"
                        }
                      ] 
                    },
                    {
                      "linkId": "9.1.3",
                      "type": "string",
                      "definition": "http://hl7.org/fhir/StructureDefinition/Observation#Observation.code",
                      "extension": [
                        {
                          "url": "http://hl7.org/fhir/StructureDefinition/questionnaire-hidden",
                          "valueBoolean": true
                        }
                      ],
                      "initial": [
                        {
                          "valueCoding": {
                            "code": "8888",
                            "display": "dummy",
                            "system": "dummy"
                          }
                        }
                      ]
                    }
                  ]
                }
              ]
            }
```

2. Use below json as questionnaire response string.
```JSON
{
              "resourceType": "QuestionnaireResponse",
              "item": [
                {
                  "linkId": "9",
                  "item": [
                    {
                      "linkId": "9.1",
                      "item" : [
                        {
                          "linkId": "9.1.1",
                          "answer": [
                            {
                              "valueCoding": {
                                "system": "test-coding-system",
                                "code": "test-coding-code",
                                "display": "Test Coding Display"
                              }
                            }
                          ]
                        }
                      ]
                    },
                    {
                      "linkId": "9.1.3",
                      "answer": [
                        {
                          "valueCoding": {
                            "system": "dummy",
                            "code": "8888",
                            "display": "dummy"
                          }
                        }
                      ]
                    }
                  ]
                }
              ]
            }
```
3. Perform extraction using ` ResourceMapper.extract(questionnaire, questionnaireResponse)`

**Expected behavior**
Extracted observation value should have valueCoding but it throws `NullPointerException`.

@jingtang10 @aditya-07

### PR DESCRIPTION
**Pull request by vbothe23**
**IMPORTANT: All PRs must be linked to an issue (except for extremely trivial and straightforward changes).**

Fixes #1769 

**Description**
Issue:
In `addAnswerToListField` if the answerValue is of non-primitive type then the primitiveValue will be null and app crashes with NullPointerException.

Fix:
Added isPrimitiveType check to handle answerValue of primitive and non-primitive types when extracting to list in definition based extraction.

**Alternative(s) considered**
No

**Type**
Choose one: Bug fix

**Screenshots (if applicable)**

**Checklist**
- [x] I have read and acknowledged the [Code of conduct](https://github.com/google/android-fhir/blob/master/docs/CODE_OF_CONDUCT.md).
- [x] I have read the [Contributing](https://github.com/google/android-fhir/wiki/Contributing) page.
- [x] I have signed the Google [Individual CLA](https://cla.developers.google.com/about/google-individual), or I am covered by my company's [Corporate CLA](https://cla.developers.google.com/about/google-corporate ).
- [x] I have discussed my proposed solution with code owners in the linked issue(s) and we have agreed upon the general approach.
- [x] I have run `./gradlew spotlessApply` and `./gradlew spotlessCheck` to check my code follows the style guide of this project.
- [x] I have run `./gradlew check` and `./gradlew connectedCheck` to test my changes locally.
- [x] I have built and run the demo app(s) to verify my change fixes the issue and/or does not break the demo app(s).

### CODE REVIEW

File: datacapture/src/test/java/com/google/android/fhir/datacapture/mapping/ResourceMapperTest.kt [hunk 1]
```diff
@@ -491,6 +491,128 @@ class ResourceMapperTest {
...
+      iParser.parseResource(QuestionnaireResponse::class.java, questionnaireResponseJson)
+        as QuestionnaireResponse
+
+    val observation =
+      ResourceMapper.extract(uriTestQuestionnaire, uriTestQuestionnaireResponse)
+        .entry
+        .single()
+        .resource as Observation
+
+    observation.value
```
Comment (joiskash) [inline datacapture/src/test/java/com/google/android/fhir/datacapture/mapping/ResourceMapperTest.kt:611]: Remove this line since it is not doing anything

**Review by joiskash — CHANGES_REQUESTED** [review]
Looks good! Congratulations on your first PR, I left a small comment please have a look at it

File: datacapture/src/test/java/com/google/android/fhir/datacapture/mapping/ResourceMapperTest.kt (hunk 1 above)
Comment (vbothe23) [reply datacapture/src/test/java/com/google/android/fhir/datacapture/mapping/ResourceMapperTest.kt]: okay.

File: datacapture/src/main/java/com/google/android/fhir/datacapture/mapping/ResourceMapper.kt [hunk 2]
```diff
@@ -671,12 +671,18 @@ private fun setFieldElementValue(base: Base, field: Field, answerValue: Base) {
...
-  base.javaClass
-    .getMethod(
-      "add${field.name.capitalize(Locale.ROOT)}",
-      answerValue.first().primitiveValue().javaClass
-    )
-    .let { method -> answerValue.forEach { method.invoke(base, it.primitiveValue()) } }
+  if (answerValue.first().isPrimitive) {
+    base.javaClass
+      .getMethod(
+        "add${field.name.capitalize(Locale.ROOT)}",
```
Comment (omarismail94) [inline datacapture/src/main/java/com/google/android/fhir/datacapture/mapping/ResourceMapper.kt:677]: capitalize is a deprecated call. Can you replace it with `replaceFirstChar`?

File: datacapture/src/main/java/com/google/android/fhir/datacapture/mapping/ResourceMapper.kt [hunk 3]
```diff
@@ -671,12 +671,18 @@ private fun setFieldElementValue(base: Base, field: Field, answerValue: Base) {
 }
 
 private fun addAnswerToListField(base: Base, field: Field, answerValue: List<Base>) {
-  base.javaClass
-    .getMethod(
-      "add${field.name.capitalize(Locale.ROOT)}",
-      answerValue.first().primitiveValue().javaClass
-    )
-    .let { method -> answerValue.forEach { method.invoke(base, it.primitiveValue()) } }
+  if (answerValue.first().isPrimitive) {
```
Comment (omarismail94) [inline datacapture/src/main/java/com/google/android/fhir/datacapture/mapping/ResourceMapper.kt:674]: is there a way we can get rid of this if/else, and just have one method call that works on both primitives and non-primitives?

For example:
```
answerValue.first().fhirType().capitalize(Locale.ROOT).javaClass
```

I experimented with this quickly and it seems like it works. Can you further test?

**Review by omarismail94 — COMMENTED** [review]
Nice change @vbothe23 ! Left a few comments

File: datacapture/src/test/java/com/google/android/fhir/datacapture/mapping/ResourceMapperTest.kt [hunk 4]
```diff
@@ -491,6 +491,125 @@ class ResourceMapperTest {
...
+                    }
+                  ]
+                }
+              ]
+            }
+          ]
+        }
+      """.trimIndent()
+
+    val iParser: IParser = FhirContext.forR4().newJsonParser()
```
[the remark this answers is not in the document]
Comment (vbothe23) [reply datacapture/src/test/java/com/google/android/fhir/datacapture/mapping/ResourceMapperTest.kt:596]: okay

File: datacapture/src/main/java/com/google/android/fhir/datacapture/mapping/ResourceMapper.kt (hunk 2 above)
Comment (vbothe23) [reply datacapture/src/main/java/com/google/android/fhir/datacapture/mapping/ResourceMapper.kt:677]: Sure

File: datacapture/src/main/java/com/google/android/fhir/datacapture/mapping/ResourceMapper.kt (hunk 3 above)
Comment (vbothe23) [reply datacapture/src/main/java/com/google/android/fhir/datacapture/mapping/ResourceMapper.kt:674]: Sure. I will test.

**Review by joiskash — APPROVED** [review]
LGTM!

**Comment by omarismail94** [comment]
LGTM, and will let @jingtang10  give the final approval!

Jing - I reviewed the PR and the suggestion I made was to use a method, fhirType(), that works on both primitives and non-primitives

File: datacapture/src/main/java/com/google/android/fhir/datacapture/mapping/ResourceMapper.kt [hunk 5]
```diff
@@ -673,8 +673,8 @@ private fun setFieldElementValue(base: Base, field: Field, answerValue: Base) {
 private fun addAnswerToListField(base: Base, field: Field, answerValue: List<Base>) {
   base.javaClass
     .getMethod(
-      "add${field.name.capitalize(Locale.ROOT)}",
-      answerValue.first().primitiveValue().javaClass
+      "add${field.name.replaceFirstChar(Char::uppercase)}",
+      answerValue.first().fhirType().replaceFirstChar(Char::uppercase).javaClass
     )
     .let { method -> answerValue.forEach { method.invoke(base, it.primitiveValue()) } }
```
Comment (jingtang10) [inline datacapture/src/main/java/com/google/android/fhir/datacapture/mapping/ResourceMapper.kt:679]: this line is still takeing the `it.primitiveValue()` from the answer value. can you tell me who this is working?

**Comment by jingtang10** [comment]
@joiskash @omarismail94 @vbothe23 can you please help addressing my above comments in a follow up cl?

**Comment by joiskash** [comment]
Sure @vbothe23  will do this ASAP.

File: datacapture/src/test/java/com/google/android/fhir/datacapture/mapping/ResourceMapperTest.kt [hunk 6]
```diff
@@ -491,6 +490,123 @@ class ResourceMapperTest {
...
+                "initial": [
+                  {
+                    "valueCoding": {
+                      "code": "8888",
+                      "display": "dummy",
+                      "system": "dummy"
+                    }
+                  }
+                ]
+              }
```
[the remark this answers is not in the document]
Comment (vbothe23) [reply datacapture/src/test/java/com/google/android/fhir/datacapture/mapping/ResourceMapperTest.kt:544]: okay

File: datacapture/src/test/java/com/google/android/fhir/datacapture/mapping/ResourceMapperTest.kt [hunk 7]
```diff
@@ -491,6 +490,123 @@ class ResourceMapperTest {
...
+                  "answer": [
+                    {
+                      "valueCoding": {
+                        "system": "dummy",
+                        "code": "8888",
+                        "display": "dummy"
+                      }
+                    }
+                  ]
+                }
```
[the remark this answers is not in the document]
Comment (vbothe23) [reply datacapture/src/test/java/com/google/android/fhir/datacapture/mapping/ResourceMapperTest.kt:588]: okay

File: datacapture/src/test/java/com/google/android/fhir/datacapture/mapping/ResourceMapperTest.kt [hunk 8]
```diff
@@ -491,6 +490,123 @@ class ResourceMapperTest {
...
+  @Test
+  fun `extract() should extract list of non primitive values`() = runBlocking {
+    @Language("JSON")
+    val questionnaireJson =
+      """
+      {
+        "resourceType": "Questionnaire",
+        "item": [
+          {
+            "linkId": "9",
```
[the remark this answers is not in the document]
Comment (vbothe23) [reply datacapture/src/test/java/com/google/android/fhir/datacapture/mapping/ResourceMapperTest.kt:502]: Sure

File: datacapture/src/main/java/com/google/android/fhir/datacapture/mapping/ResourceMapper.kt (hunk 5 above)
Comment (vbothe23) [reply datacapture/src/main/java/com/google/android/fhir/datacapture/mapping/ResourceMapper.kt:679]: Here for non primitive values, `getMethod()` function will throw `NoSuchMethodException` which intern call `updateListFieldWithAnswer()`. UpdateListFieldWithAnswer() function will set the answer.

Comment (jingtang10) [reply datacapture/src/main/java/com/google/android/fhir/datacapture/mapping/ResourceMapper.kt:679]: actually nvm i think this code is actually clear enough.

thanks for the reply.

**Pull request merged**